kernel/src/table_features/column_mapping.rs

```diff
@@ -56,3 +59,129 @@ impl AsRef<str> for ColumnMappingMode {
+        let annotation = "delta.columnMapping.physicalName";
+        match (self.mode, field.metadata.get(annotation)) {
+            // Both Id and Name modes require a physical name; None mode requires no physical name.
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, Some(MetadataValue::String(_))) => {}
+            (ColumnMappingMode::None, None) => {}
+            (ColumnMappingMode::None, Some(_)) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is not enabled but field '{}' is annotated with {}",
+                    ColumnName::new(&self.path),
+                    annotation
+                )));
+            }
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, _) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is enabled but field '{}' lacks the string annotation {}",
+                    ColumnName::new(&self.path),
+                    annotation
+                )));
+            }
+        }
+
+        let annotation = "delta.columnMapping.id";
+        match (self.mode, field.metadata.get(annotation)) {
+            // Both Id and Name modes require a field ID; None mode requires no field id.
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, Some(MetadataValue::Number(_))) => {}
+            (ColumnMappingMode::None, None) => {}
+            (ColumnMappingMode::None, Some(_)) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is not enabled but field '{}' is annotated with {}",
+                    ColumnName::new(&self.path),
+                    annotation
+                )));
+            }
+            (ColumnMappingMode::Name | ColumnMappingMode::Id, _) => {
+                self.err = Some(Error::invalid_column_mapping_mode(format!(
+                    "Column mapping is enabled but field '{}' lacks the numeric annotation {}",
+                    ColumnName::new(&self.path),
+                    annotation
+                )));
+            }
+        }
+    }
+}
+impl SchemaTransform for ValidateColumnMappings {
+    // Override array element and map key/value for better error messages
+    fn transform_array_element<'a>(&mut self, etype: &'a DataType) -> Option<Cow<'a, DataType>> {
+        self.transform_inner_type(etype, "<array element>")
+    }
+    fn transform_map_key<'a>(&mut self, ktype: &'a DataType) -> Option<Cow<'a, DataType>> {
+        self.transform_inner_type(ktype, "<map key>")
+    }
+    fn transform_map_value<'a>(&mut self, vtype: &'a DataType) -> Option<Cow<'a, DataType>> {
+        self.transform_inner_type(vtype, "<map value>")
+    }
+
+    fn transform_struct_field<'a>(
+        &mut self,
+        field: &'a StructField,
+    ) -> Option<Cow<'a, StructField>> {
+        if self.err.is_none() {
+            self.path.push(field.name.clone());
```

Comment:
It's a shame to have to clone here. The lifetimes get... messy if we try and fix this because the `SchemaTransform` trait specifies lifetime at the function level, but we need guarantees on the lifetime of  the`ValidateColumnMappings` struct itself. I think we can't fix it without hoisting the lifetime to be part of the `SchemaTransform` trait, which would mean we're always asserting lifetimes over the entire cycle of the transforming struct. I think that might be actually be good. Basically we're asserting that everything will live as long as the transformer, which is reasonable. It is a bigger change though, so happy to discuss and take as a follow-up if we think it's a good idea.

Reply:
I tried it -- actually not that hard and this is the PR to try it in if we would do it at all -- but compilation fails with:
```
error[E0277]: the trait bound `std::string::String: std::convert::From<&&str>` is not satisfied
   --> kernel/src/table_features/column_mapping.rs:134:21
    |
134 |                     ColumnName::new(&self.path),
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<&&str>` is not implemented for `std::string::String`, which is required by `expressions::column_names::ColumnName: std::iter::FromIterator<&&str>`
    |
    = help: the following other types implement trait `std::convert::From<T>`:
              `std::string::String` implements `std::convert::From<&mut str>`
              `std::string::String` implements `std::convert::From<&std::string::String>`
              `std::string::String` implements `std::convert::From<&str>`
              `std::string::String` implements `std::convert::From<char>`
              `std::string::String` implements `std::convert::From<hdfs_native::security::digest::Qop>`
              `std::string::String` implements `std::convert::From<object_store::path::Path>`
              `std::string::String` implements `std::convert::From<std::borrow::Cow<'a, str>>`
              `std::string::String` implements `std::convert::From<std::boxed::Box<str>>`
            and 4 others
    = note: required for `&&str` to implement `std::convert::Into<std::string::String>`
note: required for `expressions::column_names::ColumnName` to implement `std::iter::FromIterator<&&str>`
   --> kernel/src/expressions/column_names.rs:104:23
    |
104 | impl<A: Into<String>> FromIterator<A> for ColumnName {
    |         ------------  ^^^^^^^^^^^^^^^     ^^^^^^^^^^
    |         |
    |         unsatisfied trait bound introduced here
note: required by a bound in `expressions::column_names::ColumnName::new`
   --> kernel/src/expressions/column_names.rs:20:15
    |
18  |     pub fn new<A>(iter: impl IntoIterator<Item = A>) -> Self
    |            --- required by a bound in this associated function
19  |     where
20  |         Self: FromIterator<A>,
    |               ^^^^^^^^^^^^^^^ required by this bound in `ColumnName::new`

```

Reply:
Since the place we need to construct `ColumnName`s is only in the error path, could we just be gross and do something like:
```rust
ColumnName::new(self.path.iter().map(|s| s.to_string())
```

Reply:
I ended up defining a closure:
```rust
let column_name = || ColumnName::new(self.path.iter().copied());
```
.. and then updated the half dozen use sites to just invoke `column_name()`.

Reply:
applied in https://github.com/delta-io/delta-kernel-rs/pull/543